packages/flame_3d/shaders/spatial_material.frag

```diff
@@ -9,48 +15,171 @@ out vec4 outColor;
+  vec3 diffuse
+) {
+  vec3 lightDirVec = lightPos - fragPosition;
+  vec3 lightDir = normalize(lightDirVec);
+  float distance = length(lightDirVec) + EPSILON;
+  vec3 halfwayDir = normalize(viewDir + lightDir);
+
+  float attenuation = lightIntensity / (distance * distance);
+  vec3 radiance = lightColor * attenuation;
+
+  // cook-torrance brdf
+  float ndf = distributionGGX(normal, halfwayDir, material.roughness);
+  float g = geometrySmith(normal, viewDir, lightDir, material.roughness);
+  vec3 f = fresnelSchlick(max(dot(halfwayDir, viewDir), 0.0), diffuse);
+
+  vec3 kS = f; // reflection/specular fraction
+  vec3 kD = (vec3(1.0) - kS) * (1.0 - material.metallic); // refraction/diffuse fraction
+
+  vec3 numerator = ndf * g * f;
+  float denominator = 4.0 * max(dot(normal, viewDir), 0.0) * max(dot(normal, lightDir), 0.0) + EPSILON;
+  vec3 specular = numerator / denominator;
+
+  // add to outgoing radiance Lo
+  float NdotL = max(dot(normal, lightDir), 0.0);
+  return (kD * baseColor / PI + specular) * radiance * NdotL;
 }
 
 void main() {
-    vec3 viewDir = normalize(camera.position - fragPosition);
-    vec3 lightDir = normalize(light.position - fragPosition);
-    vec3 halfwayDir = normalize(viewDir + lightDir);
+  vec3 normal = normalize(fragNormal);
+  vec3 viewDir = normalize(camera.position - fragPosition);
+
+  vec3 baseColor = material.albedoColor;
+  baseColor *= texture(albedoTexture, fragTexCoord).rgb;
+
+  vec3 baseAmbient = vec3(0.03) * baseColor * ambientLight.color * ambientLight.intensity;
+  vec3 ao = vec3(1.0); // white - no ambient occlusion for now
+  vec3 ambient = baseAmbient * baseColor * ao;
+
+  vec3 f0 = vec3(0.04);
+  vec3 diffuse = mix(f0, baseColor, material.metallic);
+
+  vec3 lo = vec3(0.0);
+
+  // for(uint i = 0; i < lightsInfo.numLights; ++i) {
+  if (lightsInfo.numLights > 0) {
+    vec3 light0Pos = light0.position;
+    vec3 light0Color = light0.color;
+    float light0Intensity = light0.intensity;
+
+    lo += processLight(light0Pos, light0Color, light0Intensity, baseColor, normal, viewDir, diffuse);
+  }
+
+  if (lightsInfo.numLights > 1) {
+    vec3 light1Pos = light1.position;
+    vec3 light1Color = light1.color;
+    float light1Intensity = light1.intensity;
+
+    lo += processLight(light1Pos, light1Color, light1Intensity, baseColor, normal, viewDir, diffuse);
+  }
 
-    vec3 normal = normalize(fragNormal);
-    float NdotV = max(dot(normal, viewDir), 0.0);
-    float fresnel = SchlickGGX(NdotV, material.roughness);
+  if (lightsInfo.numLights > 2) {
+    vec3 light2Pos = light2.position;
+    vec3 light2Color = light2.color;
+    float light2Intensity = light2.intensity;
 
-    float NdotL = max(dot(normal, lightDir), 0.0);
-    float NdotH = max(dot(normal, halfwayDir), 0.0);
-    float specular = SchlickGGX(NdotL, material.roughness) * SchlickGGX(NdotH, material.roughness);
+    lo += processLight(light2Pos, light2Color, light2Intensity, baseColor, normal, viewDir, diffuse);
+  }
+  // }
```

Comment:
Can be removed right?

Reply:
I'm just leaving to kinda explain what the code should look like once we had arrays - but can remove if you prefer